### PR TITLE
HPCC-24630 heuristic to turn off remoteKeyedLookup if narrow key

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@mckellyln - please review

File: thorlcr/activities/keyedjoin/thkeyedjoin.cpp [hunk 1]
```diff
@@ -404,6 +404,30 @@ class CKeyedJoinMaster : public CMasterActivity
                 if (totalIndexParts)
                 {
                     indexMap.map(*this, indexFile, keyHasTlk, getOptBool("allLocalIndexParts"));
+                    if (remoteKeyedLookup && !getOptBool(THOROPT_FORCE_REMOTE_KEYED_LOOKUP))
+                    {
+                        /* heuristic to determin how many nodes in this cluster, have had parts
```
Comment (mckellyln) [inline thorlcr/activities/keyedjoin/thkeyedjoin.cpp:409]: determine

File: thorlcr/activities/keyedjoin/thkeyedjoin.cpp [hunk 2]
```diff
@@ -404,6 +404,30 @@ class CKeyedJoinMaster : public CMasterActivity
...
+                         */ 
+                        unsigned nodesWithMappedParts = 0;
+                        unsigned groupSize = queryDfsGroup().ordinality();
+                        for (unsigned s=0; s<groupSize; s++)
+                        {
+                            std::vector<unsigned> &parts = indexMap.querySlaveParts(s);
+                            if (parts.size())
+                                nodesWithMappedParts++;
+                        }
+                        if ((nodesWithMappedParts * 2) < groupSize) // i.e. if less than 50% nodes have any parts mapped
```
Comment (mckellyln) [inline thorlcr/activities/keyedjoin/thkeyedjoin.cpp:423]: Not worth it, but an option to set this threshold (as a percentage perhaps) might one day be useful ?

**Comment by mckellyln** [comment]
@jakesmith looks good, just one spelling fix

File: thorlcr/activities/keyedjoin/thkeyedjoin.cpp (hunk 2 above)
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoin.cpp:423]: too many options as is ;) I think on/off is enough in general (can be forced either way already globally or per KJ)
The heuristic here is probably slightly overkill already, i.e. vs just spotting non totally local key and disabling.

Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoin.cpp:423]: Fixed typo'

**Comment by jakesmith** [comment]
@richardkchapman - please merge when smoketest passed.

**Pull request merged**